pallets/avn/src/tests/session_handler_tests.rs

```diff
@@ -1,167 +1,272 @@
-        );
-
-        change_validators_empty();
-
-        assert!(AVN::validators() == vec![]);
-    });
-}
+    #[test]
+    fn if_no_changes_between_sessions_then_avn_knows_same_collators()
+    {
+        let mut ext = setup_initial_collators();
 
-#[test]
-// * changed is false: keys list has not changed
-fn keys_populated_correctly_new_session_with_no_change() {
-    let mut ext = ExtBuilder::build_default().with_validators().as_externality();
-    ext.execute_with(|| {
-        assert!(
-            AVN::validators() ==
-                vec![
-                    Validator { account_id: 1, key: UintAuthorityId(1) },
-                    Validator { account_id: 2, key: UintAuthorityId(2) },
-                    Validator { account_id: 3, key: UintAuthorityId(3) }
-                ]
-        );
-
-        advance_session_no_validators_change();
-
-        assert!(
-            AVN::validators() ==
-                vec![
-                    Validator { account_id: 1, key: UintAuthorityId(1) },
-                    Validator { account_id: 2, key: UintAuthorityId(2) },
-                    Validator { account_id: 3, key: UintAuthorityId(3) }
-                ]
-        );
-    });
-}
+        ext.execute_with(|| {
+            let initial_collators = avn_known_collators();
+
+            advance_session();
+
+            let current_collators = avn_known_collators();
+
+            advance_session();
+
+            assert_eq!(initial_collators, current_collators);
+            assert_eq!(current_collators, vec![
+                TestAccount::derive_validator(1),
+                TestAccount::derive_validator(2),
+                TestAccount::derive_validator(3),
+            ]);
+        });
+    }
+
+    mod when_new_candidate_registers {
+        use super::*;
+
+    #[test]
+        fn then_no_change_visible_in_following_session(){
+            let mut ext = setup_initial_collators();
+            let added_valditator = TestAccount::derive_validator(4);
+
+            ext.execute_with(|| {
+                let initial_collators = avn_known_collators();
+                add_collator_candidate(added_valditator.account_id, 4);
+
+                advance_session();
+
+                let final_collators = avn_known_collators();
+                assert_eq!(initial_collators, final_collators);
+            })
+        }
+
+    #[test]
+        fn then_avn_knows_collator_after_two_sessions(){
+            let mut ext = setup_initial_collators();
+            let added_valditator = TestAccount::derive_validator(4);
+
+            ext.execute_with(|| {
+                add_collator_candidate(added_valditator.account_id, 4);
+                advance_session();
+                advance_session();
+
+                let final_collators = avn_known_collators();
+
+                assert_eq!(final_collators,
+                    vec![
+                    TestAccount::derive_validator(3),
+                    TestAccount::derive_validator(4),
```

Comment:
Why is the new collator in this position in the vector? Is this random? If so, the test will fail eventually and we should look for something else. If not, then what is the logic to place 2nd and not 4th?

Reply:
Not sure what defines the order of the collators in the vector but to fix this, a method to sort the collators has been implemented.